### PR TITLE
feature - design favorite view

### PR DESCRIPTION
**Pull request by at-dungduong**

### CODE REVIEW

File: MyApp/Library/Ext/UIImageViewExt.swift [hunk 1]
```diff
@@ -5,25 +5,25 @@
 //  Created by Khoa Vo.T.A. on 3/1/18.
 //  Copyright © 2018 Asian Tech Co., Ltd. All rights reserved.
 //
-import Foundation
```
Comment (at-daiho) [inline MyApp/Library/Ext/UIImageViewExt.swift]: WARNING - There should be a space after //

File: MyApp/Library/Ext/UIImageViewExt.swift [hunk 2]
```diff
@@ -5,25 +5,25 @@
...
-import UIKit
```
Comment (at-daiho) [inline MyApp/Library/Ext/UIImageViewExt.swift]: WARNING - There should be a space after //

File: MyApp/Library/Ext/UIImageViewExt.swift [hunk 3]
```diff
@@ -5,25 +5,25 @@
...
-import SDWebImage
-
-extension UIImageView {
```
Comment (at-daiho) [inline MyApp/Library/Ext/UIImageViewExt.swift]: WARNING - There should be a space after //

**Comment by at-daiho** [comment]
Total linter violations in pull request: 3
Serious: 0
Warnings: 3

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: Cell/HeaderCell.swift [hunk 4]
```diff
@@ -8,33 +8,110 @@
...
+            case .failure(_):
+                SVProgressHUD.dismiss()
+            }
+        }
+    }
+
+    func configButton() {
+        guard let viewModel = viewModel else { return }
+        if viewModel.checkIsFavorite() {
+            likeButton.setImage(#imageLiteral(resourceName: "icons8-heart-filled"), for: .normal)
```
Comment (at-tuanmai2) [inline Cell/HeaderCell.swift]: set button select

File: Cell/HeaderCell.swift [hunk 5]
```diff
@@ -8,33 +8,110 @@
...
+            likeButton.setImage(#imageLiteral(resourceName: "icons8-heart-detail"), for: .normal)
+        }
+    }
+
+    private func addFavorite() {
+        guard let viewModel = viewModel else { return }
+        viewModel.addFavoritePromotion { [] (result) in
+            switch result {
+            case .success:
+                SVProgressHUD.dismiss()
```
Comment (at-tuanmai2) [inline Cell/HeaderCell.swift]: nt

File: Cell/HeaderCell.swift [hunk 6]
```diff
@@ -8,33 +8,110 @@
...
+    }
+
+    private func removeFavorite() {
+        guard let viewModel = viewModel else { return }
+        viewModel.removeFavoritePromotion { [] (result) in
+            switch result {
+            case .success:
+                viewModel.deleteOneFavoriteItem(id: viewModel.id)
+                SVProgressHUD.dismiss()
+            case .failure(_): break
```
Comment (at-tuanmai2) [inline Cell/HeaderCell.swift]: dismiss?

File: Cell/HeaderCell.swift [hunk 7]
```diff
@@ -8,33 +8,110 @@
...
+    }
+
     @IBAction func backButtonTouchUpInside(_ sender: Any) {
         AppDelegate.shared.changeViewTabBar()
     }
+
+    @IBAction func likeButtonTouchUpInside(_ sender: Any) {
+        if likeButton.currentImage == #imageLiteral(resourceName: "icons8-heart-detail") {
+            SVProgressHUD.show()
+            likeButton.setImage(#imageLiteral(resourceName: "icons8-heart-filled"), for: .normal)
```
Comment (at-tuanmai2) [inline Cell/HeaderCell.swift]: use button select

File: Demo/Login/Login/Tabbar/FavoriteCell.swift [hunk 8]
```diff
@@ -10,25 +10,48 @@ import UIKit
...
+    var viewModel: FavoriteCellViewModel? {
+        didSet {
+            updateView()
+        }
+    }
     
     override func awakeFromNib() {
         super.awakeFromNib()
+        configUI()
+        customLabel()
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoriteCell.swift]: UI?

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 9]
```diff
@@ -7,59 +7,139 @@
...
+    }
+    
     override func viewDidLoad() {
         super.viewDidLoad()
         title = App.String.favorite
-        let deleteButoon = UIBarButtonItem(image: #imageLiteral(resourceName: "icons8-empty-trash-40"), style: .plain, target: self, action: nil)
-        navigationItem.rightBarButtonItem = deleteButoon
+        let deleteButton = UIBarButtonItem(image: #imageLiteral(resourceName: "icons8-remove"), style: .plain, target: self, action: #selector(deleteAll))
+        navigationItem.rightBarButtonItem = deleteButton
+        navigationItem.rightBarButtonItem?.tintColor = #colorLiteral(red: 0.8378158212, green: 0.1756822765, blue: 0.232298106, alpha: 1)
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: defile color hết dùm anh cái! dùng đi dùng lại mà

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 10]
```diff
@@ -7,59 +7,139 @@
...
+    override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
-        viewModel.createFavoriteItems()
+        SVProgressHUD.show()
+        viewModel.getItems { [weak self] (result) in
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                this.updateView()
+                SVProgressHUD.dismiss()
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: dismiss ra ngoài

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 11]
```diff
@@ -7,59 +7,139 @@
...
+            case .failure(let error):
+                SVProgressHUD.dismiss()
+                this.alert(error: error)
+            }
+        }
+        viewModel.fetchData()
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: setupData

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 12]
```diff
@@ -7,59 +7,139 @@
...
+        viewModel.delegate = self
+    }
+    
+    func setupData() {
+        viewModel.getItems { [weak self] (result) in
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                SVProgressHUD.dismiss()
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: dismiss?

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 13]
```diff
@@ -7,59 +7,139 @@
...
+    }
+    
+    @objc private func deleteAll() {
+        SVProgressHUD.show()
+        viewModel.removeAllFavoritePromotion { [weak self] (result) in
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                this.viewModel.deleteFavoriteItem()
+                this.updateView()
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: dismiss

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 14]
```diff
@@ -7,59 +7,139 @@
...
+            guard let id = viewModel.items?[indexPath.row].id else { return }
+            viewModel.id = id
+            viewModel.deleteOneFavoriteItem(id: id)
+            viewModel.removeFavoritePromotion { [weak self] (result) in
+                guard let this = self else { return }
+                switch result {
+                case .success:
+                    SVProgressHUD.dismiss()
+                case .failure(let error):
+                    this.alert(error: error)
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: dissmiss

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 15]
```diff
@@ -7,59 +7,139 @@
...
-    }
-    
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
         return viewModel.numberOfItems(inSection: section)
     }
     
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
-        guard let cell = tableView.dequeueReusableCell(withIdentifier: "FavoriteCell", for: indexPath) as? FavoriteCell else { fatalError(App.String.errorCell)}
-        cell.viewModel = viewModel.viewModelForItem(at: indexPath) as? HomeCellViewModel
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: "FavoriteCell", for: indexPath) as? FavoriteCell else {
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: define

File: Demo/Login/Login/ViewController/LoginViewController.swift [hunk 16]
```diff
@@ -87,12 +87,13 @@ class LoginViewController: UIViewController, MVVM.View {
         case .success:
             SVProgressHUD.show()
             viewModel.processLogin(completion: { [weak self] (result) in
-                SVProgressHUD.popActivity()
                 guard let this = self else { return }
                 switch result {
                 case .success:
                     AppDelegate.shared.changeViewTabBar()
+                    SVProgressHUD.dismiss()
```
Comment (at-tuanmai2) [inline Demo/Login/Login/ViewController/LoginViewController.swift]: nt

File: Demo/Login/Login/Tabbar/FavoritesViewController.swift [hunk 17]
```diff
@@ -7,59 +7,139 @@
...
+    @objc private func deleteAll() {
+        SVProgressHUD.show()
+        viewModel.removeAllFavoritePromotion { [weak self] (result) in
+            guard let this = self else { return }
+            switch result {
+            case .success:
+                this.viewModel.deleteFavoriteItem()
+                this.updateView()
+                SVProgressHUD.dismiss()
+                this.alert(title: "Delete All?", msg: "Do You Want Delete All Your Favorites", buttons: ["OK"], handler: nil)
```
Comment (at-tuanmai2) [inline Demo/Login/Login/Tabbar/FavoritesViewController.swift]: show ở đây làm gì? delete xong rồi mới hỏi là sao?

**Pull request merged**